File: packages/svg/test/render.test.ts
import { describe, expect, test } from 'bun:test';
import { avatar } from '../src/index';
import { renderSpec } from '../src/renderers/index';
import { grainStyle, facesStyle, brutalistStyle, terminalStyle, pixelStyle } from '@sigil/core';

describe('avatar()', () => {
	test('returns valid SVG string', () => {
		const svg = avatar('Alice Chen');
		expect(svg).toContain('<svg');
		expect(svg).toContain('</svg>');
		expect(svg).toContain('xmlns="http://www.w3.org/2000/svg"');
	});

	test('respects size option', () => {
		const svg = avatar('Alice', { size: 120 });
		expect(svg).toContain('width="120"');
		expect(svg).toContain('height="120"');
	});

	test('respects style option', () => {
		const grain = avatar('Alice', { style: 'grain' });
		const pixel = avatar('Alice', { style: 'pixel' });
		expect(grain).not.toBe(pixel);
	});

	test('is deterministic', () => {
		const a = avatar('Balazs Otakomaiya', { style: 'faces' });
		const b = avatar('Balazs Otakomaiya', { style: 'faces' });
		expect(a).toBe(b);
	});

	test('different names produce different output', () => {
		const a = avatar('Alice');
		const b = avatar('Bob');
		expect(a).not.toBe(b);
	});
});

describe('renderSpec()', () => {
	test('renders grain spec', () => {
		const spec = grainStyle('Alice');
		const svg = renderSpec(spec, 80);
		expect(svg).toContain('<svg');
		expect(svg).toContain('Instrument Serif');
	});

	test('renders faces spec', () => {
		const spec = facesStyle('Alice');
		const svg = renderSpec(spec, 80);
		expect(svg).toContain('<svg');
		expect(svg).toContain('ellipse');
	});

	test('renders brutalist spec', () => {
		const spec = brutalistStyle('Alice');
		const svg = renderSpec(spec, 80);
		expect(svg).toContain('<svg');
		expect(svg).toContain('Space Mono');
	});

	test('renders terminal spec', () => {
		const spec = terminalStyle('Alice');
		const svg = renderSpec(spec, 80);
		expect(svg).toContain('<svg');
		expect(svg).toContain('IBM Plex Mono');
	});

	test('renders pixel spec', () => {
		const spec = pixelStyle('Alice');
		const svg = renderSpec(spec, 80);
		expect(svg).toContain('<svg');
		expect(svg).toContain('crispEdges');
	});

	test('all styles produce valid SVG for various names', () => {
		const names = ['Alice Chen', 'Bob M', 'Clara', 'Fatima', 'X'];
		const styles = [grainStyle, facesStyle, brutalistStyle, terminalStyle, pixelStyle] as const;

		for (const name of names) {
			for (const styleFn of styles) {
				const spec = styleFn(name);
				const svg = renderSpec(spec, 80);
				expect(svg).toContain('<svg');
				expect(svg).toContain('</svg>');
			}
		}
	});
});

File: packages/dart/test/render_test.dart
import 'package:sigil/sigil.dart';
import 'package:test/test.dart';

void main() {
  const names = ['Alice Chen', 'Bob M', 'Fatima', 'George Tanaka', 'X'];

  group('avatar()', () {
    test('returns a non-empty SVG string', () {
      for (final name in names) {
        final svg = avatar(name);
        expect(svg, startsWith('<svg'));
        expect(svg, contains('</svg>'));
        expect(svg.length, greaterThan(100));
      }
    });

    test('is deterministic', () {
      for (final name in names) {
        expect(avatar(name), equals(avatar(name)));
      }
    });

    test('different names produce different SVGs', () {
      expect(avatar('Alice'), isNot(equals(avatar('Bob'))));
    });

    test('all 8 styles render without error', () {
      for (final style in StyleId.values) {
        for (final name in names) {
          final svg = avatar(name, style: style);
          expect(svg, startsWith('<svg'), reason: '$style / $name');
          expect(svg, contains('</svg>'), reason: '$style / $name');
        }
      }
    });

    test('size parameter is reflected in SVG', () {
      final svg120 = avatar('Alice', size: 120);
      expect(svg120, contains('width="120"'));
      expect(svg120, contains('height="120"'));
    });
  });

  group('renderSpec()', () {
    test('grain renders SVG with clip path and filter', () {
      final spec = resolve('Alice Chen', StyleId.grain) as GrainSpec;
      final svg = renderSpec(spec, 80);
      expect(svg, contains('feTurbulence'));
      expect(svg, contains('feBlend'));
      expect(svg, contains(spec.palette.bg));
      expect(svg, contains(spec.initials));
    });

    test('brutalist renders SVG with mask', () {
      final spec = resolve('Bob', StyleId.brutalist) as BrutalistSpec;
      final svg = renderSpec(spec, 80);
      expect(svg, contains('<mask'));
      expect(svg, contains(spec.initials));
    });

    test('pixel renders SVG with crispEdges', () {
      final spec = resolve('Alice', StyleId.pixel) as PixelSpec;
      final svg = renderSpec(spec, 80);
      expect(svg, contains('crispEdges'));
    });

    test('borderRadius square omits rx', () {
      final svg = avatar('Alice', borderRadius: BorderRadius.square);
      expect(svg, contains('rx="0"'));
    });

    test('borderRadius round produces rx="50"', () {
      final svg = avatar('Alice', borderRadius: BorderRadius.round);
      expect(svg, contains('rx="50"'));
    });
  });
}
